core/TypeErrorDiagnostics.cc

```diff
@@ -55,19 +55,46 @@ void TypeErrorDiagnostics::maybeAutocorrect(const GlobalState &gs, ErrorBuilder
     if (gs.suggestUnsafe.has_value()) {
         e.replaceWith(fmt::format("Wrap in `{}`", *gs.suggestUnsafe), loc, "{}({})", *gs.suggestUnsafe,
                       loc.source(gs).value());
-    } else {
-        auto withoutNil = Types::approximateSubtract(gs, actualType, Types::nilClass());
-        if (!withoutNil.isBottom() &&
-            Types::isSubTypeUnderConstraint(gs, constr, withoutNil, expectedType, UntypedMode::AlwaysCompatible)) {
-            e.replaceWith("Wrap in `T.must`", loc, "T.must({})", loc.source(gs).value());
-        } else if (Types::isSubTypeUnderConstraint(gs, constr, expectedType, Types::Boolean(),
-                                                   UntypedMode::AlwaysCompatible)) {
-            if (core::isa_type<ClassType>(actualType)) {
-                auto classSymbol = core::cast_type_nonnull<ClassType>(actualType).symbol;
-                if (classSymbol.exists() && classSymbol.data(gs)->owner == core::Symbols::root() &&
-                    classSymbol.data(gs)->name == core::Names::Constants::Boolean()) {
-                    e.replaceWith("Prepend `!!`", loc, "!!({})", loc.source(gs).value());
-                }
+        return;
+    }
+
+    // If we expected an array of T, but got an array of nilable T, add an autocorrect
+    // to add .compact.
+    auto arrayOfUntyped = Types::arrayOfUntyped();
+    if (Types::isSubType(gs, expectedType, arrayOfUntyped) && Types::isSubType(gs, actualType, arrayOfUntyped)) {
+        auto *expected = core::cast_type<AppliedType>(expectedType);
+        auto *actual = core::cast_type<AppliedType>(actualType);
+        if (!expected || !actual) {
+            return;
+        }
+
+        if (expected->targs.size() != 1 || actual->targs.size() != 1) {
+            return;
+        }
+
+        auto &expectedElementType = expected->targs[0];
+        auto &actualElementType = actual->targs[0];
+        auto actualWithoutNil = Types::approximateSubtract(gs, actualElementType, Types::nilClass());
```

Comment:
This is a small thing but because you know the symbol for NilClass, you can call dropSubtypesOf directly, skipping the allocation (approximateSubtract will just immediately unwrap the TypePtr and call that function instead)